### PR TITLE
ignore canonical mime header key TT-116

### PR DESCRIPTION
**Pull request by gernest**
## Description
This PR add  `ignore_canonical_mime_header_key` config option. When this is set to true. All headers that are supplied to tyk will be used as they are when proxying without  trying to normalise them.

For instance header key `X-CertificateOuid` will passed as `X-Certificateouid` by default, if you want to preserve the header key setting `ignore_canonical_mime_header_key` to true will pass the key as `X-CertificateOuid`

__NOTE__
This is global option and it applies at gw level so there is no per api option for this.
## Related Issue

https://tyktech.atlassian.net/browse/TT-116?atlOrigin=eyJpIjoiYWYzZjU2MmM2ODEwNDlkMjk5ZmQxZmQxNzczZWNjODkiLCJwIjoiaiJ9

## How This Has Been Tested
Has unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

File: gateway/batch_requests.go [hunk 1]
```diff
@@ -84,7 +83,7 @@ func (b *BatchRequestHandler) DecodeBatchRequest(r *http.Request) (BatchRequestS
 
 func (b *BatchRequestHandler) ConstructRequests(batchRequest BatchRequestStructure, unsafe bool) ([]*http.Request, error) {
 	requestSet := []*http.Request{}
-
+	ignoreCanonical := config.Global().IgnoreCanonicalMIMEHeaderKey
```
Comment (letzya) [inline gateway/batch_requests.go:86]: @gernest out of interest  - why do we need this local variable?

**Comment by letzya** [comment]
@gernest You wrote `if you want to preserve the header key setting ignore_canonical_mime_header_key to true will pass the key as X-CertificateOuid` - doesn't it mean that we will break backward compatibility?

**Comment by gernest** [comment]
> doesn't it mean that we will break backward compatibility?
@letzya  It depends, because most servers normalises the headers upon receiving them, except for the the few which this PR covers.

 So if `ignore_canonical_mime_header_key` is set, headers managed by the GW will still be normalised, only headers provided by `configs`,`api defs`, `plugins` `external middlewares` etc will be added as they are and written upstream unmodified.

**Comment by letzya** [comment]
@gernest yes of course! thanks

**Review by matiasinsaurralde — APPROVED** [review]
Looks good, if we fix the conflicts, I would be happy to merge it.

**Review by tbuchaillot — APPROVED** [review]
LGTM

**Comment by matiasinsaurralde** [comment]
Merged to `release-3.1` (cf34c4007b11a7d9272450e6c6e4c6caf889392c), `release-3-lts` (bdc9191494b0df6675ba552255dec07c0edb08ae), `master` (ead180e01608c9e32eb3e596587b4349077fc1d0).

**Comment by mcandre** [comment]
Guys, the sys/unix merge left in merge banners, breaking compilation.

https://github.com/TykTechnologies/tyk/pull/3249/files#diff-6460c9a4dad7f5e2885177393db5376fb949773f0cc6ccd9505e674d463a42c6

**Comment by joshblakeley** [comment]
@matiasinsaurralde ☝️

**Comment by matiasinsaurralde** [comment]
@mcandre Are you building plugins using our Docker env.?

**Comment by mcandre** [comment]
@matiasinsaurralde I think the tag I am using (v3.0.1) is okay. I've only noticed master so far. Because I had been working around plugin / reflect glitches with manual `go get`... commands, followed by `git checkout <tag> && go install ./...`, in a dependency resolution shell script.

At the moment, I am replacing `go get`... with `if [ ! -e <dir> ]; then git clone`..., so that the broken master code doesn't halt the dependency resolution script.

**Pull request merged**